app/models/user.rb

```diff
@@ -216,10 +220,127 @@ def send_devise_notification(notification, *args)
+      # i18n-tasks-use t('account.email_language.name.es')
+      # i18n-tasks-use t('account.email_language.name.fr')
+      I18n.t("account.email_language.name.#{email_language}")
+    else
+      I18n.t('account.email_language.name.en')
+    end
+  end
+
+  def visible_email_addresses
+    email_addresses.filter do |email_address|
+      email_address.confirmed? || !email_address.confirmation_period_expired?
+    end
+  end
+
+  def lockout_time_expiration
+    second_factor_locked_at + lockout_period
+  end
+
+  def active_identity_for(service_provider)
+    active_identities.find_by(service_provider: service_provider.issuer)
+  end
+
+  def active_or_pending_profile
+    active_profile || pending_profile
+  end
+
+  def pending_profile_requires_verification?
+    return false if pending_profile.blank?
+    return true if identity_not_verified?
+    return false if active_profile_newer_than_pending_profile?
+    true
+  end
+
+  def identity_not_verified?
+    !identity_verified?
+  end
+
+  def identity_verified?(service_provider: nil)
+    active_profile.present? && !reproof_for_irs?(service_provider: service_provider)
+  end
+
+  def reproof_for_irs?(service_provider:)
+    return false unless service_provider&.irs_attempts_api_enabled
+    return false unless active_profile.present?
+    !active_profile.initiating_service_provider&.irs_attempts_api_enabled
+  end
+
+  def active_profile_newer_than_pending_profile?
+    active_profile.activated_at >= pending_profile.created_at
+  end
+
+    # This user's most recently activated profile that has also been deactivated
+  # due to a password reset, or nil if there is no such profile
+  def password_reset_profile
+    profile = profiles.where.not(activated_at: nil).order(activated_at: :desc).first
+    profile if profile&.password_reset?
+  end
+
+  def qrcode(otp_secret_key)
+    options = {
+      issuer: APP_NAME,
+      otp_secret_key: otp_secret_key,
+      digits: TwoFactorAuthenticatable::OTP_LENGTH,
+      interval: IdentityConfig.store.totp_code_interval,
+    }
+    url = ROTP::TOTP.new(otp_secret_key, options).provisioning_uri(
+      EmailContext.new(self).last_sign_in_email_address.email,
+    )
+    qrcode = RQRCode::QRCode.new(url)
+    qrcode.as_png(size: 240).to_data_url
+  end
+
+  def locked_out?
+    second_factor_locked_at.present? && !lockout_period_expired?
+  end
+
+  def no_longer_locked_out?
+    second_factor_locked_at.present? && lockout_period_expired?
+  end
+
+  def recent_events
+    events = Event.where(user_id: id).order('created_at DESC').limit(MAX_RECENT_EVENTS).
+      map(&:decorate)
+    (events + identity_events).sort_by(&:happened_at).reverse
+  end
+
+  def identity_events
+    identities.includes(:service_provider_record).order('last_authenticated_at DESC')
+  end
+
+  def recent_devices
+    @recent_devices ||= devices.order(last_used_at: :desc).limit(MAX_RECENT_DEVICES).
+      map(&:decorate)
+  end
+
+  def devices?
```

Comment:
should we rename this to `recent_devices?` so is less likely to be confused with the `devices` association?

Reply:
Done in https://github.com/18F/identity-idp/pull/8204/commits/72b3e7dfa2634bac10cb813f2c71f5fbd42d5e22